core/src/main/java/org/apache/brooklyn/core/network/OnPublicNetworkEnricher.java

```diff
@@ -0,0 +1,496 @@
+                            new Object[] {OnPublicNetworkEnricher.this, event.getSensor().getName(), event.getValue(), entity});
+                    tryTransform((AttributeSensor<?>)event.getSensor());
+                }});
+        }
+        if (mapMatching.isPresent()) {
+            Sensor<?> wildcardSensor = null;
+            subscriptions().subscribe(entity, wildcardSensor, new SensorEventListener<Object>() {
+                @Override public void onEvent(SensorEvent<Object> event) {
+                    if (mapMatching.get().apply(event.getSensor())) {
+                        LOG.debug("{} attempting transformations, triggered by sensor-event {}->{}, to {}", 
+                                new Object[] {OnPublicNetworkEnricher.this, event.getSensor().getName(), event.getValue(), entity});
+                        tryTransform((AttributeSensor<?>)event.getSensor());
+                    }
+                }});
+        }
+
+        tryTransformAll();
+    }
+
+    @Override
+    public void destroy() {
+        try {
+            if (pfmListener != null) {
+                getPortForwardManager().removeAssociationListener(pfmListener);
+            }
+        } finally {
+            super.destroy();
+        }
+    }
+
+    protected void tryTransformAll() {
+        if (!isRunning()) {
+            return;
+        }
+        Maybe<MachineLocation> machine = getMachine();
+        if (machine.isAbsent()) {
+            return;
+        }
+        for (AttributeSensor<?> sensor : sensors) {
+            try {
+                tryTransform(machine.get(), sensor);
+            } catch (Exception e) {
+                // TODO Avoid repeated logging
+                Exceptions.propagateIfFatal(e);
+                LOG.warn("Problem transforming sensor "+sensor+" of "+entity, e);
+            }
+        }
+        if (mapMatching.isPresent()) {
+            for (Sensor<?> sensor : entity.getEntityType().getSensors()) {
+                if (sensor instanceof AttributeSensor && mapMatching.get().apply(sensor)) {
+                    try {
+                        tryTransform(machine.get(), (AttributeSensor<?>)sensor);
+                    } catch (Exception e) {
+                        // TODO Avoid repeated logging
+                        Exceptions.propagateIfFatal(e);
+                        LOG.warn("Problem transforming sensor "+sensor+" of "+entity, e);
+                    }
+                }
+            }
+        }
+    }
+
+    protected void tryTransform(AttributeSensor<?> sensor) {
+        if (!isRunning()) {
+            return;
+        }
+        Maybe<MachineLocation> machine = getMachine();
+        if (machine.isAbsent()) {
+            return;
+        }
+        tryTransform(machine.get(), sensor);
+    }
+    
+    protected void tryTransform(MachineLocation machine, AttributeSensor<?> sensor) {
+        Object sensorVal = entity.sensors().get(sensor);
+        if (sensorVal == null) {
+            return;
```

Comment:
I wondered about that. It depends why the port-mapping was removed from the PFM. If it's because the entity is stopping, then the enricher shouldn't bother trying to keep up with that. But if it's because the the mappings are more dynamic, then it would make sense to remove it.

My reasoning was that we don't delete the main.uri sensor for a Tomcat entity when the tomcat process stops, so we can similarly leave these sensors as-is.

But like I said, if things are a lot more dynamic for DNAT rules added/removed on-the-fly then we should revisit.